pex/pep_376.py

```diff
@@ -0,0 +1,430 @@
+    if PY2:
+        from hashlib import _hash as _Hash
+    else:
+        from hashlib import _Hash
+    from typing import Callable, Dict, Iterable, Iterator, Optional, Protocol, Tuple, Union
+
+    import attr  # vendor:skip
+
+    class CSVWriter(Protocol):
+        def writerow(self, row):
+            # type: (Iterable[Union[str, int]]) -> None
+            pass
+
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class Digest(object):
+    algorithm = attr.ib()  # type: str
+    encoded_hash = attr.ib()  # type: str
+
+    def new_hasher(self):
+        # type: () -> _Hash
+        return hashlib.new(self.algorithm)
+
+
+@attr.s(frozen=True)
+class Hash(object):
+    value = attr.ib()  # type: str
+
+    def __str__(self):
+        # type: () -> str
+        return self.value
+
+    def parse(self):
+        # type: () -> Digest
+        algorithm, encoded_hash = self.value.split("=", 1)
+        return Digest(algorithm=algorithm, encoded_hash=encoded_hash)
+
+
+@attr.s(frozen=True)
+class InstalledFile(object):
+    """The record of a single installed file from a PEP 376 RECORD file.
+
+    See: https://www.python.org/dev/peps/pep-0376/#record
+    """
+
+    path = attr.ib()  # type: str
+    hash = attr.ib(default=None)  # type: Optional[Hash]
+    size = attr.ib(default=None)  # type: Optional[int]
+
+
+class InstallationScheme(Enum["InstallationScheme.Value"]):
+    """Represents the Pip installation scheme used for installing a wheel.
+
+    For more about installation schemes, see:
+        https://docs.python.org/3/install/index.html#alternate-installation
+
+    N.B.: Pex only uses the --target scheme but all schemes are represented for documentation
+    purposes. Notably, Pex _could_ change to using the --prefix scheme with changes to its runtime
+    `sys.path` adjustments in order to afford less hackery when reading the RECORD.
+    """
+
+    class Value(Enum.Value):
+        pass
+
+    PREFIX = Value("--prefix")
+    ROOT = Value("--root")
+    TARGET = Value("--target")
+    USER = Value("--user")
+
+
+class RecordError(Exception):
+    pass
+
+
+class ReinstallError(RecordError):
+    """Indicates an error re-installing an installed distribution."""
+
+
+class RecordNotFoundError(RecordError):
+    """Indicates a distribution's RECORD metadata could not be found."""
+
+
+@attr.s(frozen=True)
+class Record(object):
+    """Represents the PEP-376 RECORD of an installed wheel.
+
+    See: https://www.python.org/dev/peps/pep-0376/#record
+    """
+
+    @classmethod
+    def read(
+        cls,
+        lines,  # type: Union[FileInput[str], Iterator[str]]
+        exclude=None,  # type: Optional[Callable[[str], bool]]
+    ):
+        # type: (...) -> Iterator[InstalledFile]
+
+        # The RECORD is a csv file with the path to each installed file in the 1st column.
+        # See: https://www.python.org/dev/peps/pep-0376/#record
+        for line, (path, fingerprint, file_size) in enumerate(
```

Comment:
Given the janky state of some PyPI artifacts, I'm guessing malformed RECORD files may exist, and this unpacking may fail? Possibly worth checking the length of the parsed line before attempting to unpack and skipping instead of failing on any messed up lines?

Reply:
Thankfully Pex is immune from all but vendored Pip`s quirks here because Pex uses it to do the install; i.e. it writes the RECORD.